framework/src/transfers/MultiAppReporterTransfer.C

```diff
@@ -133,15 +158,63 @@ MultiAppReporterTransfer::executeFromMultiapp()
   if (!hasFromMultiApp())
     return;
 
-  unsigned int ind = _subapp_index != std::numeric_limits<unsigned int>::max() ? _subapp_index : 0;
-  if (getFromMultiApp()->hasLocalApp(ind) &&
-      (!hasToMultiApp() || getToMultiApp()->hasLocalApp(ind)))
-    for (unsigned int n = 0; n < _from_reporter_names.size(); ++n)
-      transferReporter(_from_reporter_names[n],
-                       _to_reporter_names[n],
-                       getFromMultiApp()->appProblemBase(ind),
-                       hasToMultiApp() ? getToMultiApp()->appProblemBase(ind)
-                                       : getFromMultiApp()->problemBase());
+  // subapp indices to perform transfers on
+  std::vector<unsigned int> indices;
+  if (_distribute_reporter_vector)
+  {
+    // If distributing, resize the indices vector to the number of global apps
+    indices.resize(getFromMultiApp()->numGlobalApps());
+    std::iota(indices.begin(), indices.end(), 0);
+  }
+  else if (_subapp_index == std::numeric_limits<unsigned int>::max())
+  {
+    // if _subapp_index not set indices is set to 0
+    indices = {0};
+  }
+  else
+    // set indices to specific _subapp_index
+    indices = {_subapp_index};
+
+  if (_distribute_reporter_vector)
+    for (unsigned int n : make_range(_to_reporter_names.size()))
+    {
+      // Clear all vector reporters and resize to the number of subapps.
+      // The summing process later will make sure the reporter values are
+      // consistent across the processors.
+      auto size = getFromMultiApp()->numGlobalApps();
+      clearVectorReporter(_to_reporter_names[n], getFromMultiApp()->problemBase());
+      resizeReporter(_to_reporter_names[n], getFromMultiApp()->problemBase(), size);
+    }
+
+  for (const auto & ind : indices)
```

Comment:
```suggestion
  for (const auto ind : indices)
```
no need for references for trivially copyable types